ts/receiver/multidevice.ts

```diff
@@ -0,0 +1,70 @@
+import { removeFromCache } from './cache';
+import { EnvelopePlus } from './types';
+
+async function unpairingRequestIsLegit(source: string, ourPubKey: string) {
+  const { textsecure, storage, lokiFileServerAPI } = window;
+
+  const isSecondary = textsecure.storage.get('isSecondaryDevice');
+  if (!isSecondary) {
+    return false;
+  }
+  const primaryPubKey = storage.get('primaryDevicePubKey');
+  // TODO: allow unpairing from any paired device?
+  if (source !== primaryPubKey) {
+    return false;
+  }
+
+  const primaryMapping = await lokiFileServerAPI.getUserDeviceMapping(
+    primaryPubKey
+  );
+
+  // If we don't have a mapping on the primary then we have been unlinked
+  if (!primaryMapping) {
+    return true;
+  }
+
+  // We expect the primary device to have updated its mapping
+  // before sending the unpairing request
+  const found = primaryMapping.authorisations.find(
+    (authorisation: any) => authorisation.secondaryDevicePubKey === ourPubKey
+  );
+
+  // our pubkey should NOT be in the primary device mapping
+  return !found;
+}
+
+async function clearAppAndRestart() {
+  // remove our device mapping annotations from file server
+  await window.lokiFileServerAPI.clearOurDeviceMappingAnnotations();
+  // Delete the account and restart
+  try {
+    await window.Signal.Logs.deleteAll();
+    await window.Signal.Data.removeAll();
```

Comment:
these calls to Data can be imported directly.
```
import { removeAll } from 'js/modules/data.js';
```
just keep in mind doing:
```
import * as Data from 'js/modules.data'
```
will make it so stubbing those data functions in tests won't be possible. So if this code isn't going to be tested yet we might want to use the latter for now.